From 252d33874fbc194c1af7b342581d4656a04cf146 Mon Sep 17 00:00:00 2001
From: Adam Ford <aford173@gmail.com>
Date: Wed, 29 Apr 2020 07:17:14 -0500
Subject: [PATCH 22/25] arm64: dts: beacon-imx8mm: Fix Camera power on Beta
 hardware

The alpha boards had an issue where the camera power was left on during
suspend/power off, because the power was active low.

In beta hardware, the power is active high, so when the GPIO pins
go low the camera is disabled.

To help maintain backwards comatibility with older hardware, a new
device tree called beacon-imx8mm-kit-alpha which undoes the polarity
change, but the user will have to specify this device tree in the
bootloader if they have alpha boards.

Fixes: PHOENIX8-55
Signed-off-by: Adam Ford <aford173@gmail.com>
---
 .../freescale/beacon-imx8mm-baseboard.dtsi    | 13 +++++++++
 .../dts/freescale/beacon-imx8mm-kit-alpha.dts | 28 +++++++++++++++++++
 2 files changed, 41 insertions(+)
 create mode 100644 arch/arm64/boot/dts/freescale/beacon-imx8mm-kit-alpha.dts

diff --git a/arch/arm64/boot/dts/freescale/beacon-imx8mm-baseboard.dtsi b/arch/arm64/boot/dts/freescale/beacon-imx8mm-baseboard.dtsi
index 17a02fddfcb3..48bba64db8e6 100644
--- a/arch/arm64/boot/dts/freescale/beacon-imx8mm-baseboard.dtsi
+++ b/arch/arm64/boot/dts/freescale/beacon-imx8mm-baseboard.dtsi
@@ -66,6 +66,16 @@
 		};
 	};
 
+	reg_camera: regulator-camera {
+		compatible = "regulator-fixed";
+		regulator-name = "mipi_pwr_nen";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&pca6416_1 0 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		startup-delay-us = <100000>;
+	};
+
 	reg_pcie_pwr_en: regulator-pci_pwr_en {
 		compatible = "regulator-fixed";
 		regulator-name = "pci_pwr_en";
@@ -349,6 +359,9 @@
 		assigned-clock-parents = <&clk IMX8MM_CLK_24M>;
 		assigned-clock-rates = <0>, <24000000>;
 		csi_id = <0>;
+		DOVDD-supply = <&reg_camera>;
+		AVDD-supply = <&reg_camera>;
+		DVDD-supply = <&reg_camera>;
 		pwn-gpios = <&gpio1 7 GPIO_ACTIVE_HIGH>;
 		mclk = <24000000>;
 		mclk_source = <0>;
diff --git a/arch/arm64/boot/dts/freescale/beacon-imx8mm-kit-alpha.dts b/arch/arm64/boot/dts/freescale/beacon-imx8mm-kit-alpha.dts
new file mode 100644
index 000000000000..765006facb44
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/beacon-imx8mm-kit-alpha.dts
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright 2020 Beacon EmbeddedWorks
+ */
+
+/dts-v1/;
+
+#include "fsl-imx8mm.dtsi"
+#include "beacon-imx8mm-som.dtsi"
+#include "beacon-imx8mm-baseboard.dtsi"
+#include "beacon-imx8mm-lvds.dtsi"
+
+/ {
+	model = "beacon EmbeddedWorks i.MX8M mini Developent Kit";
+	compatible = "fsl,imx8mm";
+
+	chosen {
+		stdout-path = &uart2;
+	};
+};
+
+/* For Alpha boards, camera power is active low */
+&reg_camera {
+	gpio = <&pca6416_1 0 GPIO_ACTIVE_LOW>;
+	/delete-property/ enable-active-high;
+};
+
+
-- 
2.17.1

